### PR TITLE
Fix for "Route not found" issue in custom controllers using pagerfanta

### PR DESCRIPTION
**Pull request by wimvds**
| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | 
| Related issues/PRs | 
| License | MIT
| Documentation PR |

#### What's in this PR?

A basic route caching implementation to make getRouteByName work.

#### Why?

While integrating (already existing) custom controllers that used pagerfanta for paging data, we always got a "Route not found" exception. The root cause seemed to be that the getRouteByName function was not defined in the ContentRouteProvider. This PR just caches the route collection for the current request, so the route can be found again when the pagerfanta widget tries to generate it (with paging parameters added).

### CODE REVIEW

**Review by wachterjohannes — APPROVED** [review]
LGTM and this specific usecase seems to work also for me. but I am not 100% sure if this is a good idea, because that solves only a quite specific problem.

Maybe that could be a part of solving the problem of generating routes by name. @danrot @chirimoya what do you think? It would be also OK for me to merge this as it is.

**Comment by wimvds** [comment]
I'm fully aware that it only fixes this specific use case (ie. anything that uses the route of the current request) and we could also override the provider in our own projects for now. But I think this could help a lot of users that are integrating existing code. No hard feelings if it doesn't get merged though. Still loving Sulu :+1:.

BTW If you were to decide to merge it, would it be ok if I add the same fix to the Sulu\Component\CustomUrl\Routing\CustomUrlRouteProvider as well?

**Comment by wachterjohannes** [comment]
@wimvds if we merge it, it would be no problem to add it to all places we need it :) one thing we talked internal today is that we should add a comment to the code why we have added this and which usecase will be fixed by this code. Would you do that?

**Comment by wimvds** [comment]
Sure, I'll do it later today or tomorrow.

**Review by danrot — APPROVED** [review]
For me it is also ok to merge. Better support a single use case than no usecase at all.

**Review by wachterjohannes — APPROVED** [review]
@danrot can be merged in my opinion

**Comment by wimvds** [comment]
I noticed that there's still an issue with this : even though it works on the website, the preview in the back-end does not work. I'll try to find out why.

**Comment by wachterjohannes** [comment]
@wimvds will you fix this preview bug also?

**Comment by wimvds** [comment]
I'm looking into it. Will keep you updated.

**Comment by wimvds** [comment]
Nope, I kind of lost track of this issue (we did the easy "fix" and decided to not show the pager on previews). Will have a look at it today, IIRC I was stuck with the routing in the preview kernel.

**Comment by wimvds** [comment]
Ran into dependency hell yesterday when trying to upgrade the project we had the issue with (it uses Sylius as well ...). So I decided to create a dummy project (https://github.com/wimvds/pager-sulu-demo) to find a better way to resolve the issue.

**Comment by wimvds** [comment]
The root cause of the issue seems to be that the ```sulu_route.routing.router``` is not added in the preview kernel. Trying to figure out why.

**Comment by wimvds** [comment]
Ok, the root cause seemed to be that the Symfony routing is completely bypassed in the PreviewRenderer. This results in no _route being defined in the request info, and thus causes the "None of the chained routers were able to generate route: Route '' not found, Route 'Symfony\Cmf\Component\Routing\ProviderBasedGenerator' not found" exception. 

So I added a _route in the getByEntity result of the PageRouteDefaultsProvider and added a (dirty) hack to the PreviewRenderer to inject the route (with the actual path of the content).

File: src/Sulu/Component/CustomUrl/Routing/CustomUrlRouteProvider.php [hunk 1]
```diff
@@ -180,4 +189,24 @@ private function decodePathInfo($pathInfo)
     {
         return rawurldecode($pathInfo);
     }
+
+    /**
+     * @param RouteCollection $collection
+     */
+    public function setRouteCollection(RouteCollection $collection)
```
Comment (danrot) [inline src/Sulu/Component/CustomUrl/Routing/CustomUrlRouteProvider.php]: I don't think it is a good idea to introduce this method, if you only need it for tests... Then I would even prefer to set it via reflections in the test. If we introduce this method people might start using it for whatever reason...

File: src/Sulu/Bundle/WebsiteBundle/Routing/ContentRouteProvider.php [hunk 2]
```diff
@@ -282,4 +291,24 @@ private function decodePathInfo($pathInfo)
 
         return '/' . ltrim(rawurldecode($pathInfo), '/');
     }
+
+    /**
+     * @param RouteCollection $collection
+     */
+    public function setRouteCollection(RouteCollection $collection)
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Routing/ContentRouteProvider.php]: Same as below, would even prefer reflections in the test if this is the only reason.

**Review by danrot — CHANGES_REQUESTED** [review]
There is still a CI error... Can you also try to rebase? I guess that the reason for the circle ci error.
@wachterjohannes Can you also have another look?

**Comment by wimvds** [comment]
Done. Should I also squash the commits or leave them as is?

**Review by danrot — APPROVED** [review]

**Comment by danrot** [comment]
@wimvds I am really sorry, seem to have lost track of this PR... Is this still something you want to get merged?

**Comment by niklasnatter** [comment]
There are a lot of merge conflicts and the last commit on this was in 2017, therefore I am going to close this. Feel free to open a new PR if this is still a problem - thanks a lot for your time!

**Pull request closed without merging**